envi/archs/h8/emu.py

```diff
@@ -0,0 +1,1314 @@
+        sp = self.getRegister(reg)
+        sp -= inc
+        self.writeMemValue(sp, val, inc)
+        self.setRegister(reg, sp)
+
+    def doPop(self, inc=2, reg=REG_SP):
+        sp = self.getRegister(reg)
+        val = self.readMemValue(sp, inc)
+        self.setRegister(reg, sp+inc)
+        return val
+
+    def integerSubtraction(self, op):
+        """
+        Do the core of integer subtraction but only *return* the
+        resulting value rather than assigning it.
+        (allows cmp and sub to use the same code)
+        """
+        # Src op gets sign extended to dst
+        #FIXME account for same operand with zero result for PDE
+        ssize = op.opers[0].tsize
+        dsize = op.opers[1].tsize
+        src = self.getOperValue(op, 0)
+        dst = self.getOperValue(op, 1)
+
+        if dst == None or src == None:
+            self.undefFlags()
+            return None
+
+        return self.intSubBase(dst, dsize, src, ssize)
+
+    def intSubBase(self, src1, dsize, src2, ssize):
+        # So we can either do a BUNCH of crazyness with xor and shifting to
+        # get the necessary flags here, *or* we can just do both a signed and
+        # unsigned sub and use the results.
+
+        usrc = e_bits.unsigned(src1, ssize)
+        udst = e_bits.unsigned(src2, dsize)
+
+        ssrc = e_bits.signed(src1, ssize)
+        sdst = e_bits.signed(src2, dsize)
+
+        ures = udst - usrc
+        sres = sdst - ssrc
+
+        self.setFlag(CCR_H, e_bits.is_signed_half_carry(ures, dsize, udst))
+        self.setFlag(CCR_C, e_bits.is_unsigned_carry(ures, dsize))
+        self.setFlag(CCR_Z, not ures)
+        self.setFlag(CCR_N, e_bits.is_signed(ures, dsize))
+        self.setFlag(CCR_V, e_bits.is_signed_overflow(sres, dsize))
+
+        return ures
+
+
+    def logicalAnd(self, op):
+        src1 = self.getOperValue(op, 0)
+        src2 = self.getOperValue(op, 1)
+
+        # PDE
+        if src1 == None or src2 == None:
+            self.undefFlags()
+            self.setOperValue(op, 1, None)
+            return
+
+        res = src1 & src2
+
+        return res
+
+    def i_and(self, op):
+        dsize = op.opers[1].tsize
+        res = self.logicalAnd(op)
+        self.setOperValue(op, 1, res)
+       
+        self.setFlag(CCR_Z, not res)
+        self.setFlag(CCR_N, e_bits.is_signed(res, dsize))
+        self.setFlag(CCR_V, 0)
+
+    def i_andc(self, op):
+        res = self.logicalAnd(op)
+        self.setOperValue(op, 1, res)
+       
+
+
+    def i_band(self, op):
+        C = self.getFlag(CCR_C)
+        bit = self.getOperValue(op, 0)
+        val = self.getOperValue(op, 1)
+
+        val >>= bit
+        val &= C
+
+        self.setFlag(CCR_C, val)
+
+    def i_bra(self, op):
+        nextva = self.getOperValue(op, 0)
+        return nextva
+
+    def i_brn(self, op):
+        pass
+
+    def i_bhi(self, op):
+        if not (self.getFlag(CCR_C) == 0 or self.getFlag(CCR_Z) == 0):  return
```

Comment:
stylistically, lets keep these as returns on the next line ( for consistency if nothing else )